packages/utils/src/prune.ts

```diff
@@ -20,214 +16,118 @@ import { mapSchema } from './mapSchema';
 import { MapperKind } from './Interfaces';
 import { getRootTypes } from './rootTypes';
 
-type NamedOutputType =
-  | GraphQLObjectType
-  | GraphQLInterfaceType
-  | GraphQLUnionType
-  | GraphQLEnumType
-  | GraphQLScalarType;
-type NamedInputType = GraphQLInputObjectType | GraphQLEnumType | GraphQLScalarType;
-
-interface PruningContext {
-  schema: GraphQLSchema;
-  unusedTypes: Record<string, boolean>;
-  implementations: Record<string, Record<string, boolean>>;
-}
-
 /**
  * Prunes the provided schema, removing unused and empty types
  * @param schema The schema to prune
  * @param options Additional options for removing unused types from the schema
  */
 export function pruneSchema(schema: GraphQLSchema, options: PruneSchemaOptions = {}): GraphQLSchema {
-  const pruningContext: PruningContext = createPruningContext(schema);
+  const {
+    skipEmptyCompositeTypePruning,
+    skipEmptyUnionPruning,
+    skipPruning,
+    skipUnimplementedInterfacesPruning,
+    skipUnusedTypesPruning,
+  } = options;
+  let prunedTypes: string[] = []; // Pruned types during mapping
+  let prunedSchema: GraphQLSchema = schema;
+
+  do {
+    const visited = visit(prunedSchema, new Set<string>());
+
+    prunedTypes = [];
+
+    prunedSchema = mapSchema(prunedSchema, {
+      [MapperKind.TYPE]: type => {
+        if (!visited.has(type.name) && !isSpecifiedScalarType(type)) {
+          if (skipPruning && skipPruning(type)) {
+            return type;
+          }
+          if (isUnionType(type) || isInputObjectType(type) || isInterfaceType(type) || isObjectType(type)) {
```

Comment:
Hmmm do we not prune unused leaf types?

Reply:
I'm afraid I don't follow. These are all being pruned under this block unless told to skip. The outer `if` block checks `!visited.has(type.name)` which means it will be pruned.

Reply:
What I should have said is they the skipUnusedTypePruning seems to not be checked for leaf types.

Reply:
The only leaf type that could fall under `skipUnusedTypesPruning` (from what I can tell) is a custom scalar, which isn't caught today in the same check. I can add it.

Reply:
Although... now need to account for `skipPruning(type)` custom logic and its leaf types 😣